library/std/src/io/buffered/switchwriter.rs

```diff
@@ -0,0 +1,115 @@
+use crate::fmt::Arguments;
+use crate::io::{self, buffered::LineWriterShim, BufWriter, IoSlice, Write};
+/// Different buffering modes a writer can use
+#[unstable(feature = "stdout_switchable_buffering", issue = "78515")]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum BufferMode {
+    /// Unbuffered: forward writes directly to the underlying writer. In some
+    /// cases, a writer may buffer temporarily (for instance, when processing
+    /// a formatted write), but even in this case it will always immediately
+    /// flush.
+    None,
```

Comment:
I don't think `BufferMode::None` is a good name for a mode that still temporarily buffers things. Maybe `FlushMode::Directly` or `BufferMode::AlwaysFlush` or `BufferMode::Immediate` something?

Is there really no use case for a truly non-buffered mode? It wouldn't need to keep any buffer around in that case.

Reply:
Adding a truly non-buffered mode would also mean this buffer might be usable for Stderr as well. If it then starts in a completely unbuffered mode, the behaviour stays the same as it is now, but it'd open the possibility to enable (line) buffering on stderr too, as requested in https://github.com/rust-lang/rust/issues/64413.

Reply:
I am guessing the reasoning here is that we want to let `write_fmt` not have to allocate for each call to `write_fmt`, so it can still do multiple tiny writes and then our SwitchWriter can dump the result even if it doesn't happen to end with a newline? Is there a reason we should not try to fix that on the other end of the writing and formatting infrastructure, instead of here?

Reply:
Sorry, the reasoning for what part of it? The idea is that we already have a buffer available, so we can use it (without any extra allocation) to assemble the tiny writes which are then forwarded as a single write

Reply:
I like `Immediate` a lot, because it leaves room for the design of `write_fmt`. Changed.

Reply:
> Is there really no use case for a truly non-buffered mode? It wouldn't need to keep any buffer around in that case.

I'm struggling to imagine one. My understanding is that `stderr` is unbuffered so that any errors sent to it are shown to the user ASAP; write_fmt's design here still fulfills that goal, since it flushes immediately.

Reply:
OK, then it is what I thought it was. Still curious as to your thoughts regarding whether this is the better place to fix it, or to adjust inside `write_fmt`?

Reply:
I mean, `write_fmt` pretty much by definition needs to be defined in terms of `write_all`. All it's doing is looping over `write_all` for the various segments of the formatted write. I think in this case it makes sense to specialize it here, because we have context available to us (a buffer that would otherwise be skipped) that isn't available to generic `Write::write_fmt` calls.

Contrast to `BufWriter::write_fmt`, which has no such specialization, because the default behavior (`write_all` in a loop) is correct, since that writes directly to the buffer. Here in `SwitchWrite`, those looped calls would be forwarded directly to the underlying writer, which I am arguing should be specialized to make use of the buffer for the duration of that single formatted write.

Reply:
> I'm struggling to imagine one. 

An error encountered during printing to stderr where you still want to earlier parts to reach the output?

Reply:
You'd still get that, though, since writing to an already-allocated buffer is infallible. So you're still getting as-much-as-possible writing content to the underlying device.